### ISSUE TITLE
Support directive comment groups

### ISSUE THREAD

**Issue by tjenkinson** [eslint/eslint#13472]
**The version of ESLint you are using.**
Latest

**The problem you want to solve.**
We run eslint once on our source code and again on the build output, with different config and plugins.

Any directive comments get picked up by both the source lint and the build lint. We can disable processing directive comments completely on the build lint, but it would be useful for comments to target either the source/build lint. Some of the rules we have on the build lint have false positives so it would be useful to be able to add `/* eslint-disable-line custom-rule */` comments to disable them.

**Your take on the correct solution to problem.**
Add the ability to assign directive comments to groups, and specify the group that should be used when running `eslint`. Any comments that are not part of the provided group are ignored. The group would default to `default` and it could be a non-breaking change.

**Are you willing to submit a pull request to implement this change?**
Yes. I opened a draft (https://github.com/eslint/eslint/pull/13471) and it would be great to get some feedback before I continue and figure out how to add some tests.

**Comment by anikethsaha** [comment]
How the groups will be formed? like the group as a rule's name or category?

Can you give an example? ?

**Comment by tjenkinson** [comment]
My suggestion would be the user comes up with a group name and includes that in the config/cli option, and then use the same name in the directive comments in the code.

I put an example in the pr I linked. 

E.g if you use group name `build` in the config then

```js

// eslint[build]-disable-line no-es6
CoreUtils.find(...)

```
would match

**Comment by btmills** [comment]
This is an interesting problem, @tjenkinson, and the fact that you're already using a custom rule opens up some options. Rather than waiting for something to be implemented in ESLint core, you could address the false positives closer to where they occur in the rule. Here's one idea:

```js
// es6-features-rule-false-positive CoreUtils is not an array.
CoreUtils.find(...)
```

You could then use [`context.getSourceCode().getCommentsBefore()`](https://eslint.org/docs/developer-guide/working-with-rules#context-getsourcecode) on the `CallExpression`'s parent `ExpressionStatement` node. If the rule finds a "this is a false positive" comment, it could suppress the report.

Would that work?

**Comment by tjenkinson** [comment]
@btmills yes this is another approach we were thinking of. Thanks for the link. The api makes it look pretty straightforward :)

I think my suggestion is a bit more flexible though because it works with any rule without specific handling like this. Is it something you'd consider? I'm happy to give it a go if it makes sense.

**Comment by btmills** [comment]
Disable comments are a sometimes-necessary escape valve, and while conditional disable comments are a neat idea, I don't see us adding features to disable comments to encourage people to reach for them as anything more than a last resort. If there were some use case that couldn't be achieved another way, that might change things, but in your case you have a couple of less invasive options.

Another one I thought of since my earlier comment: if your codebase is in TypeScript, you should be able to make your rule [type-aware](https://github.com/typescript-eslint/typescript-eslint#can-we-write-rules-which-leverage-type-information) so that it only reports e.g. `.find()` on instances of `Array`, which would fix the false positive issue altogether.

If this is still something you'd like to see implemented, we can leave the issue open to see if anyone else on the team has ideas for other use cases and want to include this.

**Comment by tjenkinson** [comment]
Unfortunately our current use cases aren't typescript. I realised we could also add a config option to this plugin to include an allow list of variable names (`[ 'coreutils' ]`).

>If this is still something you'd like to see implemented, we can leave the issue open to see if anyone else on the team has ideas for other use cases and want to include this.

I'd be interested to see what others think. There is a solution for our current use case but it's not feasible as soon as we wanted to use a rule which wasn't one of ours.
